disttask/operator/operator.go

```diff
@@ -0,0 +1,94 @@
+// Copyright 2023 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package operator
+
+import (
+	"fmt"
+
+	"github.com/pingcap/tidb/resourcemanager/pool/workerpool"
+	"github.com/pingcap/tidb/resourcemanager/util"
+)
+
+// Operator is the basic operation unit in the task execution.
+type Operator interface {
+	Open() error
+	Close() error
+	Display() string
+}
+
+// AsyncOperator process the data in async way.
+//
+//		Eg: The sink of AsyncOperator op1 and the source of op2
+//	 	use the same channel, Then op2's worker will handle
+//	  	the result from op1.
+type AsyncOperator[T, R any] struct {
+	pool *workerpool.WorkerPool[T, R]
+}
+
+// NewAsyncOperator create an AsyncOperator.
+func NewAsyncOperator[T, R any](name string, workerNum int, transform func(T) R) *AsyncOperator[T, R] {
```

Comment:
previously i thought we need impl each operator separately, now it seems we have a common operator and we need to impl a different `transform`

Reply:
and it's not that flexible as each input T we generate a output R, but in some case we only generate one output from all input data.(right now, we're using a special sink as a collector)

Reply:
> and it's not that flexible as each input T we generate a output R, but in some case we only generate one output from all input data.(right now, we're using a special sink as a collector)

Do you mean that one operator should have multiple type of input？

Reply:
Let me change it to two constructors:

```go
// NewAsyncOperatorWithTransform create an AsyncOperator with a transform function.
func NewAsyncOperatorWithTransform[T, R any](name string, workerNum int, transform func(T) R) *AsyncOperator[T, R] {
	pool := workerpool.NewWorkerPool(name, util.DistTask, workerNum, newAsyncWorkerCtor(transform))
	return NewAsyncOperator(pool)
}

// NewAsyncOperator create an AsyncOperator.
func NewAsyncOperator[T, R any](pool *workerpool.WorkerPool[T, R]) *AsyncOperator[T, R] {
	return &AsyncOperator[T, R]{
		pool: pool,
	}
}
```